### PR TITLE
Typescript

### PR DESCRIPTION
**Pull request by GHaberis**
## What does this change?
This PR covers the initial conversion of our type system in dotcom-rendering from JavaScript with Flow to TypeScript. In summary the steps taken were to:

- Install Typescript as a dependency and set-up a `tsconfig.json` file
- Convert `.js` files to `.ts` or `.tsx` if they contain JSX.
- Replace flow type declarations with TypeScript equivalents
- Introduce a `.d.ts` files in `dotcom-rendering` and `frontend` which contain custom type declarations
- Install `@type` dependencies when available for our 3rd party libraries

The local dev server works as before and we still have source maps!

Built bundle sizes below:

**Typescript**
frontend.article.8353cfb100337738c608.js  48.42 kB → 13.94 kB (gzip)
frontend.server.js 126 kB       
vendor.e8f0df3d9aa90c79e582.js 139.89 kB → 45.93 kB (gzip)

**JavaScript (Master)**
frontend.article.2f98d0d515d2c034d945.js 50.22 kB → 14.09 kB (gzip)
frontend.server.js 168 kB       
vendor.3119ae5584d0490179d6.js 142.1 kB → 47.16 kB (gzip)

**Test on CODE before merging**

**TODO - these can be done in separate PRs **

- Remove `flow` dependencies and associated files
- Integrate some kind of typescript linter ([TSLint](https://palantir.github.io/tslint/) OR [typescript-eslint-parser)
](https://github.com/eslint/typescript-eslint-parser)
- Convert and types to `dotcom-rendering/libs`
- Review custom type declarations in our `.d.ts` files, I've manually written some rudimentary types for `emotion-server`, `dom-purify` and `compose-function` just to get the TSC compiler passing.

### CODE REVIEW

File: config.js [hunk 1]
```diff
@@ -4,7 +4,7 @@ const glob = promisify(require('glob'));
 
 const siteName = 'frontend';
 const getPagesForSite = () =>
-    glob('*.js', { cwd: `${siteName}/pages` }).then(paths =>
+    glob('*.?s*(x)', { cwd: `${siteName}/pages` }).then(paths =>
```
Comment (SiAdcock) [inline config.js]: Do we want this glob to be stricter?

```
'*.ts*(x)'
```

**Comment by SiAdcock** [comment]
Do we now have to import `React` into every module so we can use the `React` type? If so, it makes the [ProvidePlugin](https://github.com/guardian/dotcom-rendering/blob/master/webpack/index.js#L67-L69) pretty pointless.

File: frontend/document.tsx [hunk 2]
```diff
@@ -1,21 +1,21 @@
...
 
 type Props = {
     data: {
         page: string,
         site: string,
         body: { config?: {}, contentFields?: {} },
     },
-    Page: React.ComponentType<{}>,
+    Page: React.StatelessComponent<{data: Data}>,
+    // Page: React.ComponentType<{data: Data}>,
```
Comment (SiAdcock) [inline frontend/document.tsx]: Do we still need this?

**Comment by AWare** [comment]
We have types for `create-emotion-server` at https://github.com/guardian/manage-frontend/blob/master/app/types/create-emotion-server.d.ts but not `emotion-server`. If we want any customisations (any stylis plugins etc), we'll have to change our emotion setup to use it anyway. So I'd like to defer typing emotion properly until we've made that call.

**Pull request merged**